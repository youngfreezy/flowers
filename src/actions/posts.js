// @flow
/**
 * @module Actions/User
 * @desc User Actions
 */
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export const { getPosts } = createActions({
  [ActionTypes.GET_POSTS]: query => ({ query }),
});

export const { saveEntry } = createActions({
  [ActionTypes.SAVE_ENTRY]: entry => entry,
});
