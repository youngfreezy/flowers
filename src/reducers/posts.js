/* eslint-disable no-console */
import { parseError } from 'modules/client';
import { handleActions } from 'modules/helpers';
import sortBy from 'lodash/sortBy';
import { ActionTypes, STATUS } from 'constants/index';

export const postsState = {
  items: [],
  status: STATUS.IDLE,
};

export default {
  posts: handleActions(
    {
      [ActionTypes.GET_POSTS]: (draft, { payload }) => {
        console.log(draft);
        console.log(payload);
        console.log('THE PAYLOAD FROM GET REQUEST');
      },
      [ActionTypes.GET_POSTS_SUCCESS]: (draft, { payload }) => {
        draft.items = payload.items || [];
        draft.status = STATUS.SUCCESS;
      },
      [ActionTypes.GET_POSTS_FAILURE]: (draft, { payload }) => {
        draft.items = parseError(payload.message);
        draft.status = STATUS.ERROR;
      },
      [ActionTypes.SAVE_ENTRY]: (draft, { payload }) => {
        const newItems = sortBy(
          draft.items.filter(item => item.id !== payload.id).concat(payload),
          'id',
        );
        draft.items = newItems;
      },
    },
    postsState,
  ),
};
