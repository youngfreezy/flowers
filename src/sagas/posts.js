/*eslint-disable no-console */

/**
 * @module Sagas/Posts
 * @desc Posts
 */

import { all, call, put, takeLatest } from 'redux-saga/effects';
import { request } from 'modules/client';

import { ActionTypes } from 'constants/index';

/**
 * Get Repos
 *
 * @param {Object} action
 *
 */
export function* getPosts() {
  try {
    const response = yield call(request, '/api/posts');
    yield put({
      type: ActionTypes.GET_POSTS_SUCCESS,
      payload: { items: response.items },
    });
  } catch (err) {
    console.log(err, 'THE ERROR');
    /* istanbul ignore next */
    yield put({
      type: ActionTypes.GET_POSTS_FAILURE,
      payload: err,
    });
  }
}

export default function* root() {
  yield all([takeLatest(ActionTypes.GET_POSTS, getPosts)]);
}
