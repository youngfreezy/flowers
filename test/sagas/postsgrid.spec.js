import { expectSaga } from 'redux-saga-test-plan';

import posts, { getPosts } from 'sagas/posts';
import { ActionTypes } from 'constants/index';

jest.mock('modules/client', () => ({
  request: () => ({ items: [] }),
}));

describe('posts', () => {
  it('should have the expected watchers', done =>
    expectSaga(posts)
      .run({ silenceTimeout: true })
      .then(saga => {
        expect(saga).toMatchSnapshot();
        done();
      }));

  it('should have the posts saga', () =>
    expectSaga(getPosts)
      .put({
        type: ActionTypes.GET_POSTS_SUCCESS,
        payload: {
          items: [],
        },
      })
      .run());
});
