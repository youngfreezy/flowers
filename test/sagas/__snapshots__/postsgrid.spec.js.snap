// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`posts should have the expected watchers 1`] = `
Object {
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "GET_POSTS",
          "@@redux-saga-test-plan/json/function/getPosts",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
      "type": "FORK",
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "GET_POSTS",
      },
      "type": "TAKE",
    },
  ],
}
`;
